src/tree.rs

```diff
@@ -135,3 +138,159 @@ pub fn minimum_spanning_tree(
 
     Ok(spanning_tree)
 }
+
+/// Find balanced cut edge of the minmum spanning tree of a graph using node
+/// contraction. Assumes that the tree is connected and is a spanning tree.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. This
+///     tells retworkx/rust how to extract a numerical weight as a ``float``
+///     for edge object. Some simple examples are::
+///
+///         minimum_spanning_tree(graph, weight_fn: lambda x: 1)
+///
+///     to return a weight of 1 for all edges. Also::
+///
+///         minimum_spanning_tree(graph, weight_fn: float)
+///
+///     to cast the edge object as a float as the weight.
+/// :param float default_weight: If ``weight_fn`` isn't specified this optional
+///     float value will be used for the weight/cost of each edge.
+///
+/// :returns: A set of nodes in one half of the spanning tree
+///
+#[pyfunction]
+#[pyo3(text_signature = "(spanning_tree, pop, target_pop, epsilon)")]
+pub fn balanced_cut_edge(
+    py: Python,
+    spanning_tree: &graph::PyGraph,
+    py_pops: &PyList,
+    py_pop_target: &PyFloat,
+    py_epsilon: &PyFloat,
+) -> PyResult<Vec<(usize, Vec<usize>)>> {
+    let epsilon = py_epsilon.value();
+    let pop_target = py_pop_target.value();
+
+    let mut pops: Vec<f64> = vec![]; // not sure if the conversions are needed
+    for i in 0..py_pops.len() {
+        pops.push(py_pops.get_item(i).unwrap().extract::<f64>().unwrap());
+    }
+
+    let spanning_tree_graph = &spanning_tree.graph;
+
+    let balanced_nodes = py.allow_threads(move || {
+        let mut same_partition_tracker: Vec<Vec<usize>> =
+            vec![vec![]; spanning_tree_graph.node_count()]; // keeps track of all all the nodes on the same side of the partition
+        let mut node_queue: VecDeque<NodeIndex> = VecDeque::<NodeIndex>::new();
+        for leaf_node in spanning_tree_graph.node_indices() {
+            // todo: filter expr
+            if spanning_tree_graph.neighbors(leaf_node).count() == 1 {
+                node_queue.push_back(leaf_node);
+            }
+            same_partition_tracker[leaf_node.index()].push(leaf_node.index());
+        }
+
+        // eprintln!("leaf nodes: {}", node_queue.len());
+
+        // this process can be multithreaded, if the locking overhead isn't too high
+        // (note: locking may not even be needed given the invariants this is assumed to maintain)
+        let mut balanced_nodes: Vec<(usize, Vec<usize>)> = vec![];
+        let mut seen_nodes: Vec<bool> = vec![false; spanning_tree_graph.node_count()]; // todo: perf test this
```

Comment:
Normally for this we use a `HashSet` of `usize` (you can preallocate it with `with_capacity()` too) and then just do a lookup to see if it's been seen before.

Reply:
Yep -- I had a version of this with a HashSet -- I think I was testing this out for performance, but forgot to check if it actually improved performance (it probably doesn't). 

I think the idea I had was that it doesn't necessarily make sense to use hash tables if it's just a set of integers from 0 to n that will eventually all be in the set. Not sure if this actually makes a difference, so I was curious to see what happened.

Reply:
Sure that makes sense, it's worth testing at least.

Reply:
Ok, I think switching to `HashSet` is slightly less performant, but is worth the reduced reading headache. I've removed this rather janky code to make it easier to maintain.